File: legacy/programming-models-2x/webmvc-fn/src/test/java/com/example/demo/DemoApplicationIT.java
package com.example.demo;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.HttpStatus;

import static org.assertj.core.api.Assertions.assertThat;

class DemoApplicationIT {

    private TestRestTemplate client;

    @BeforeEach
    void setUp() {
        this.client = new TestRestTemplate();
    }

    @Test
    void getAllPosts() {
        // or use ParameterizedTypeReference to define the type
        var response = this.client.getForEntity("http://localhost:8080/", Post[].class);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody().length).isEqualTo(2);
    }

}

File: programming-models-2x/webflux-ktco-fn/src/test/kotlin/com/example/demo/DemoApplicationIT.kt
package com.example.demo


import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.web.reactive.function.client.WebClient
import reactor.kotlin.test.test

class DemoApplicationIT {

    private lateinit var client: WebClient

    private var port: Int = 8080

    @BeforeEach
    fun setup() {
        client = WebClient.create("http://localhost:$port")
    }

    @Test
    fun `get all posts`() {
        client.get()
                .uri("/")
                .accept(MediaType.APPLICATION_JSON)
                .exchangeToFlux {
                    assertThat(it.statusCode()).isEqualTo(HttpStatus.OK)
                    it.bodyToFlux(Post::class.java)
                }
                .test()
                .expectNextCount(2)
                .verifyComplete()
    }

}
